### PR TITLE
The EnsoGL Component abstraction with special dropping behavior

### PR DESCRIPTION
**Pull request by farmaazon**
[ci no changelog needed]

### Pull Request Description

In this branch:
* The workaround for cursor-not-being-updated-after-closing-searcher bug (discovered while testing #3278) is reverted.
* The proper fix was introduced: created an abstraction for EnsoGL component, which, when dropping, will not immediately drop the FRP network and model, but instead put it into the Garbage Collector. The Collector ensures, that all "component hiding" effects and events will be handled, and drops FRP network and model only after that.
* I run clippy for wasm32 target out of curiosity. There was one warning, and I fixed it on this branch.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

**Comment by MichaelMauderer** [comment]
Not an in-depth review, just a quick note: we already have a `Component` abstraction in `lib/rust/ensogl/component/gui/src/component.rs`, so I suggest we should unify them and not have two different structures implementing this.

File: lib/rust/ensogl/core/src/display/garbage.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
+//! A module with garbage [`Collector`].
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/garbage.rs]: These docs should be at least slightly better, as they are confusing right now. (Also missing "a" before "garbage").

File: lib/rust/ensogl/core/src/display/garbage.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+///
+/// # Implementation
+///
+/// Each collected component FRP network and model must be thus kept long enough until the following
+/// sequence will happen:
+/// 1. We render the scene without the component;
+/// 2. On the new rendered scene, we request the pixel value under the mouse cursor (see
+/// [`crate::display::render::passes::PixelReadPass`]).
+/// 3. The requested pixel value is received
+/// 4. The mouse events are handled (see [`crate::display::scene::SceneData::handle_mouse_events`]).
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/garbage.rs]: 1. mixed semicolons, dots, and nothing at the end of sentences.
2. Point 3 - it should be explained that "received" means that it is read into CPU memory after creating fences on GPU. Basically, this docs lacks information that the whole process can span over several frames and that later frames can get pixels from a few frames back, which is crucial here.

File: lib/rust/ensogl/core/src/display/garbage.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        let mut garbage = self.garbage.borrow_mut();
+        let objects_being_moved = std::mem::take(&mut garbage.before_pixel_update);
+        garbage.before_mouse_events.extend(objects_being_moved);
+    }
+
+    /// Mouse events handled (the point 4 in [`Collector`] docs).
+    pub fn mouse_events_handled(&self) {
+        let mut garbage = self.garbage.borrow_mut();
+        drop(std::mem::take(&mut garbage.before_mouse_events));
+    }
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/garbage.rs:84]: This is super clean!

File: lib/rust/ensogl/core/src/display/garbage.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+
+
+#[cfg(test)]
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/garbage.rs:93]: missing "Tests" section

File: lib/rust/ensogl/core/src/display/scene.rs [hunk 5]
```diff
@@ -957,6 +957,7 @@ impl SceneData {
         let new_target = PointerTarget::from_internal(self.mouse.hover_ids.get());
         let current_target = self.mouse.target.get();
         if new_target != current_target {
+            DEBUG!("Target changed");
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/scene.rs]: debug left

**Review by wdanilo — APPROVED** [review]
It's beautiful. The implementation is way nicer than I thought it will be!

**Comment by wdanilo** [comment]
> Not an in-depth review, just a quick note: we already have a `Component` abstraction in `lib/rust/ensogl/component/gui/src/component.rs`, so I suggest we should unify them and not have two different structures implementing this.

Oh, I was sure that this is the same `Component`. Hmmm, I was thinking that our main goal is to create in a single codebase as many structs with the "Component" name as possible, am I missing something? :D

**Comment by farmaazon** [comment]
> Not an in-depth review, just a quick note: we already have a `Component` abstraction in `lib/rust/ensogl/component/gui/src/component.rs`, so I suggest we should unify them and not have two different structures implementing this.

Hmm, I tried to merge those, but run into one problem: The component in "enso-component-gui" crate assumes that every component should implement application::View and provide shortcuts. Nodes so far do neither of these. @wdanilo it is ok that the Nodes (and similar components) will be registered as application views, or it's too heavy?

**Comment by wdanilo** [comment]
@farmaazon, according to our chat just a moment ago: we need to think what the component abstraction is for. Your component abstraction is the glue between FRP, model, and logic of the component. Michael's component is also an application view with shortcuts. Components will be small, like widgets on nodes (drop-down menus), they should not be application views, because it's just not their responsibility to register shortcuts. So these are 2 abstractions and one should use another one.

**Comment by farmaazon** [comment]
Applied suggestions. As some changes are non-trivial, feel free to re-review this branch

File: lib/rust/ensogl/core/src/gui/component.rs [hunk 6]
```diff
@@ -228,3 +228,94 @@ impl<T: display::Object> display::Object for ShapeView<T> {
         self.shape.display_object()
     }
 }
+
+
+
+// =================
+// === Component ===
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/gui/component.rs]: Widget

File: lib/rust/ensogl/core/src/gui/component.rs [hunk 7]
```diff
@@ -228,3 +228,94 @@ impl<T: display::Object> display::Object for ShapeView<T> {
...
+/// # Dropping Widget
+///
+/// Upon dropping Widget structure, the object will be hidden, but both FRP and Model will
+/// not be dropped at once. Instead, they will be passed to the EnsoGL's garbage collector,
+/// because handling all effects of hiding object and emitting appropriate events  (for example:
+/// the "on_hide" event of [`core::gui::component::ShapeEvents`]) may take a several frames, and
+/// we want to have FRP network alive to handle those effects. Thus, both FRP and model will
+/// be dropped only after all process of object hiding will be handled.
+#[derive(Debug)]
+pub struct Widget<Model: 'static, Frp: 'static> {
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/gui/component.rs:244]: don't use bounds in struct definition. Bounds should be used in impulse only.

File: app/gui/view/graph-editor/src/component/node.rs [hunk 8]
```diff
@@ -385,24 +386,30 @@ ensogl::define_endpoints! {
...
 #[allow(missing_docs)]
 pub struct Node {
-    pub model: Rc<NodeModel>,
-    pub frp:   Frp,
+    widget: Rc<gui::Widget<Rc<NodeModel>, Frp>>,
 }
 
 impl AsRef<Node> for Node {
-    fn as_ref(&self) -> &Self {
+    fn as_ref(&self) -> &Node {
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node.rs]: Why this change?

File: app/gui/view/graph-editor/src/component/node.rs [hunk 9]
```diff
@@ -385,24 +386,30 @@ ensogl::define_endpoints! {
...
 ///    emitted back to the right node).
 ///
-/// Currently, the solution "C" (most optimal) is implemented here.
+/// Currently, the solution "C" (nearest to optimal) is implemented here.
 #[derive(Clone, CloneRef, Debug)]
 #[allow(missing_docs)]
 pub struct Node {
-    pub model: Rc<NodeModel>,
-    pub frp:   Frp,
+    widget: Rc<gui::Widget<Rc<NodeModel>, Frp>>,
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node.rs]: to talk about rcs

**Review by wdanilo — APPROVED** [review]

**Review by MichaelMauderer — APPROVED** [review]
